### PR TITLE
[CBRD-22853] extract schema buffer

### PR DESCRIPTION
**Pull request by razvanradulescu**
http://jira.cubrid.org/browse/CBRD-22853

Add extract_schema.[ch]pp containing 
struct extract_context   : context for extract (in this PR contains only data required by unload schema to work, in the future should contain all unload context, which is now shared as external variables in unloadb.h)
class extract_output     : output wrapper for string buffer and FILE * (it will be enhanced with a method which automatically flush chunks when a size is reached - for string_buffer)
                           The preferred "append" method was the "functor" operator used also by string_buffer (this make the line changes easier).

Split trigger_manger.[ch] functionality  (tr_dump_trigger , tr_dump_selective_triggers, is_required_trigger, get_user_name ) related to schema extraction to trigger_description.[ch]pp. (maybe we should use another unit, but now contains schema related to triggers)

Split authentication.[ch] functionality (issue_grant_statement , class_grant_loop , au_export_users , au_export_grants) related to schema extraction to authentication_extract.[ch]pp.
Moved some structure definition, constants from authentication.c to authentication.h to be shared with the new files.

Replaced many (only those used in tree call of unload schema) functions which used FILE* to use an instance of extract_output.
Some functions have also been renamed (removed 'f' - no longer specialized for files) : desc_value_fprint, fprint_set, help_fprint_obj, help_fprint_describe_comment 

Major transformation of functions
extractschema  -> extract_classes_to_file  : 
 - moved the code generating the filename to specialized function
 - split the function (extract_classes_to_file , extract_triggers_to_file , extract_indexes_to_file )
 - each (.._to_file) is wrapper which uses  the new functions using the extract_output interface.
 
unloaddb  : 
 - contains more logic to generate filenames and call individual schema parts (extract_classes_to_file, extract_triggers_to_file , extract_indexes_to_file)
 - removed some global variables (output_dirname, output_file, output_prefix)
 
db_print_value (overload of db_fprint_value):
 - "specialized" code when extract_output is string_buffer to avoid copying from a temporary string_buffer memory to extract_output
 
Removed function:
desc_value_print  (not used)

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/executables/load_object.c (3 incorrect indentation(s) found)
```
    75 static void print_set (extract_output & output_ctx, DB_SET * set);
  1045 print_set (extract_output & output_ctx, DB_SET * set)
  1632 desc_value_print (extract_output & output_ctx, DB_VALUE * value)
```
#### File: src/executables/load_object.h (1 incorrect indentation(s) found)
```
    89 extern void desc_value_print (extract_output & output_ctx, DB_VALUE * value);
```
#### File: src/executables/unload_schema.c (107 incorrect indentation(s) found)
```
   130 static void emit_cycle_warning (extract_output & output_ctx);
   131 static void force_one_class (extract_output & output_ctx, DB_OBJLIST ** class_list, DB_OBJLIST ** order_list);
   132 static DB_OBJLIST *get_ordered_classes (extract_output & output_ctx, MOP * class_table);
   133 static void emit_class_owner (extract_output & output_ctx, MOP class_);
   134 static int export_serial (extract_output & output_ctx);
   135 static int emit_indexes (extract_output & output_ctx, DB_OBJLIST * classes, int has_indexes,
   136 			 DB_OBJLIST * vclass_list_has_using_index);
   138 static int emit_schema (extract_output & output_ctx, DB_OBJLIST * classes, int do_auth,
   139 			DB_OBJLIST ** vclass_list_has_using_index, EMIT_STORAGE_ORDER emit_storage_order);
   141 static DB_OBJLIST *emit_query_specs (extract_output & output_ctx, DB_OBJLIST * classes);
   142 static int emit_query_specs_has_using_index (extract_output & output_ctx, DB_OBJLIST * vclass_list_has_using_index);
   143 static bool emit_superclasses (extract_output & output_ctx, DB_OBJECT * class_, const char *class_type);
   144 static bool emit_resolutions (extract_output & output_ctx, DB_OBJECT * class_, const char *class_type);
   145 static void emit_resolution_def (extract_output & output_ctx, DB_RESOLUTION * resolution,
   146 				 RESOLUTION_QUALIFIER qualifier);
   147 static bool emit_instance_attributes (extract_output & output_ctx, DB_OBJECT * class_, const char *class_type,
   148 				      int *has_indexes, EMIT_STORAGE_ORDER storage_order);
   149 static bool emit_class_attributes (extract_output & output_ctx, DB_OBJECT * class_, const char *class_type);
   150 static bool emit_all_attributes (extract_output & output_ctx, DB_OBJECT * class_, const char *class_type,
   151 				 int *has_indexes, EMIT_STORAGE_ORDER storage_order);
```
... More than 20 lines are omitted.
#### File: src/executables/unloaddb.c (31 incorrect indentation(s) found)
```
   155   /* create here the first filename to raise error early in case output file is incorrect */
   157 			      sizeof (output_filename_schema)) != 0)
   160       goto end;
   293 				  sizeof (output_filename_schema)) != 0)
   294 	{
   295 	  util_log_write_errid (MSGCAT_UTIL_GENERIC_INVALID_ARGUMENT);
   296 	  goto end;
   297 	}
   300 				   sizeof (trigger_output_filename)) != 0)
   301 	{
   302 	  util_log_write_errid (MSGCAT_UTIL_GENERIC_INVALID_ARGUMENT);
   303 	  goto end;
   304 	}
   307 				   sizeof (indexes_output_filename)) != 0)
   308 	{
   309 	  util_log_write_errid (MSGCAT_UTIL_GENERIC_INVALID_ARGUMENT);
   310 	  goto end;
   311 	}
   318 	{
   319 	  status = 1;
```
... More than 20 lines are omitted.
#### File: src/executables/unloaddb.h (8 incorrect indentation(s) found)
```
    62 extern int extract_classes_to_file (extract_context & ctxt, const char *output_filename);
    63 extern int extract_triggers (extract_context & ctxt, extract_output & output_ctx);
    64 extern int extract_triggers_to_file (extract_context & ctxt, const char *output_filename);
    65 extern int extract_indexes_to_file (extract_context & ctxt, const char *output_filename);
    66 extern int extract_classes (extract_context & ctxt, extract_output & schema_output_ctx);
    70 				   char *output_filename_p, const size_t filename_size);
    72 				    char *output_filename_p, const size_t filename_size);
    74 				    char *output_filename_p, const size_t filename_size);
```
#### File: src/object/class_object.c (3 incorrect indentation(s) found)
```
  7442 			  att->storage_order);
  7454 	  extract_output ("\n");
  7469 			  att->storage_order);
```
#### File: src/object/object_print.c (4 incorrect indentation(s) found)
```
   142 help_print_obj (extract_output & output_ctx, MOP obj)
   164 			  cinfo.class_type, cinfo.name);
   315 			  oinfo.classname);
   717 help_print_describe_comment (extract_output & output_ctx, const char *comment)
```
#### File: src/object/object_print.h (2 incorrect indentation(s) found)
```
    50 void help_print_obj (extract_output & output_ctx, MOP obj);
    68 void help_print_describe_comment (extract_output & output_ctx, const char *comment);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by ribram** [comment]
Naming suggestion - use "print" instead of "extract" schema. It is more consistent with existing printing functionality.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/executables/unload_schema.c (1 incorrect indentation(s) found)
```
   145 static void emit_resolution_def (print_output & output_ctx, DB_RESOLUTION * resolution, RESOLUTION_QUALIFIER qualifier);
```
#### File: src/object/class_object.c (2 incorrect indentation(s) found)
```
  7442 		      att->storage_order);
  7469 		      att->storage_order);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/executables/unload_schema.c [hunk 1]
```diff
@@ -801,73 +804,81 @@ export_serial (FILE * outfp)
...
+  err_count = extract_classes (ctxt, output_ctx);
 
-  snprintf (output_filename, sizeof (output_filename) - 1, "%s/%s%s", output_dirname, output_prefix, SCHEMA_SUFFIX);
-  output_file = fopen_ex (output_filename, "w");
-  if (output_file == NULL)
-    {
-      (void) fprintf (stderr, "%s: %s.\n\n", exec_name, strerror (errno));
-      return errno;
-    }
+  fprintf (output_file, "\n");
```
Comment (eseokoh) [inline src/executables/unload_schema.c]: output_ctx?

File: src/executables/unload_schema.c [hunk 2]
```diff
@@ -950,40 +947,94 @@ extractschema (const char *exec_name, int do_auth, EMIT_STORAGE_ORDER storage_or
...
+      output_file = NULL;
+      remove (output_filename);
     }
-
-  if (vclass_list_has_using_index != NULL)
-    {
-      db_objlist_free (vclass_list_has_using_index);
+  else
+    {				/* not empty */
+      fprintf (output_file, "\n\n");
```
Comment (eseokoh) [inline src/executables/unload_schema.c]: output_ctx?

Comment (eseokoh) [reply src/executables/unload_schema.c]: same usages on `extract_triggers_to_file`.

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/object/authenticate.c (9 incorrect indentation(s) found)
```
   551 static void issue_grant_statement (print_output & output_ctx, CLASS_AUTH * auth, CLASS_GRANT * grant, int authbits);
   552 static int class_grant_loop (print_output & output_ctx, CLASS_AUTH * auth);
  2442 au_export_users (print_output & output_ctx)
  2590 		  output_ctx ("call [find_user]('%s') on class [db_user] to [auser];\n", uname);
  7482 issue_grant_statement (print_output & output_ctx, CLASS_AUTH * auth, CLASS_GRANT * grant, int authbits)
  7567 class_grant_loop (print_output & output_ctx, CLASS_AUTH * auth)
  7641 au_export_grants (print_output & output_ctx, MOP class_mop)
  7675 	      output_ctx (msgcat_message
  7676 			  (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_AUTHORIZATION, MSGCAT_AUTH_GRANT_DUMP_ERROR), uname);
```
#### File: src/object/authenticate.h (2 incorrect indentation(s) found)
```
   241 extern int au_export_users (print_output & output_ctx);
   242 extern int au_export_grants (print_output & output_ctx, MOP class_mop);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/object/authenticate.c (9 incorrect indentation(s) found)
```
   552 static void issue_grant_statement (print_output & output_ctx, CLASS_AUTH * auth, CLASS_GRANT * grant, int authbits);
   553 static int class_grant_loop (print_output & output_ctx, CLASS_AUTH * auth);
  6932 au_export_users (print_output & output_ctx)
  7080 		  output_ctx ("call [find_user]('%s') on class [db_user] to [auser];\n", uname);
  7484 issue_grant_statement (print_output & output_ctx, CLASS_AUTH * auth, CLASS_GRANT * grant, int authbits)
  7569 class_grant_loop (print_output & output_ctx, CLASS_AUTH * auth)
  7643 au_export_grants (print_output & output_ctx, MOP class_mop)
  7677 	      output_ctx (msgcat_message
  7678 			  (MSGCAT_CATALOG_CUBRID, MSGCAT_SET_AUTHORIZATION, MSGCAT_AUTH_GRANT_DUMP_ERROR), uname);
```
#### File: src/object/authenticate.h (2 incorrect indentation(s) found)
```
   241 extern int au_export_users (print_output & output_ctx);
   242 extern int au_export_grants (print_output & output_ctx, MOP class_mop);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/base/printer.hpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    void operator+= (const char ch);
+
+    template<typename... Args> inline int operator() (Args &&... args);
+};
+
+
+template<typename... Args> int print_output::operator() (Args &&... args)
+{
+  int res = m_sb (std::forward<Args> (args)...);
+  if (res < 0)
```
Comment (ribram) [inline src/base/printer.hpp]: m_sb never returns negative (it has an assert for that).

File: src/base/printer.hpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+/*
+ * print_output : interface class to print contents
+ * operator () : template variadic operator to allows printf-like syntax
+ *               it automaticaly calls virtual flush () method which needs to be implemented in specialized class
+ */
+class print_output
+{
+  private:
+    std::string context_name;
```
Comment (ribram) [inline src/base/printer.hpp]: Maybe context_name/exec_name belongs to extract_context rather than print_output?

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/executables/unload_schema.c (1 incorrect indentation(s) found)
```
   869 	  fprintf (stderr, "%s: Unknown database error occurs " "but may not be database error.\n\n", ctxt.exec_name);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Pull request merged**